Stop writing to column `user_id` of tables `profiles` and `user_filters`

H-Shay opened the pull request:
This is part 4 of a migration to drop the column `user_id` from the tables `profiles` and `user_filters`. In this stage we cease writing to the column in preparation for dropping it entirely in part 5. 

In order to continue writing to the other columns of the tables while ceasing to write to `user_id`, the `NOT NULL` constraint on this column is dropped as well.

MatMaul commented on synapse/storage/database.py:
```diff
@@ -98,6 +98,8 @@
     "event_push_summary": "event_push_summary_unique_index2",
     "receipts_linearized": "receipts_linearized_unique_index",
     "receipts_graph": "receipts_graph_unique_index",
+    "profiles": "profiles_full_user_id_key_idx",
+    "user_filters": "full_users_filters_unique_idx",
```
FTR, this is included in this PR because it is the one that begin using `full_user_id` for the upserts.

MatMaul reviewed (approved):
LGTM.

H-Shay commented:
@MatMaul thanks for reviewing - this had been sitting for awhile and after your approval I double-checked it and realized it was subject to the same error I fixed in https://github.com/matrix-org/synapse/pull/15817. I have fixed the code here, would you mind double-checking that it still looks sane? Thanks!

MatMaul commented on synapse/storage/schema/main/delta/79/01_drop_user_id_constraint_profiles.py:
```diff
@@ -0,0 +1,50 @@
+            user_id text,
+            displayname text,
+            avatar_url text,
+            UNIQUE (full_user_id),
```
Regarding your last comment @H-Shay , same pattern here I believe ?

H-Shay replied:
The `profiles` table does want these rows to be `UNIQUE` - this should be correct.

H-Shay commented:
Merged #15787 into develop, thanks for the (double) reviews @MatMaul!
